test/unit/test_shapetracker.py

```diff
@@ -227,6 +227,99 @@ def test_permute_reshape_2(self):
       new_st.append(st)
     self.sts = new_st
 
+  def test_reshaping_splitting(self):
+    self.st = CheckingShapeTracker((5,10,5,10))
+    self.st.permute((1, 0, 3, 2))
+    self.st.pad(((0,0), (0,5), (0,0), (0,5)))
+    self.st.reshape((10,2,5,10,2,5))
+    assert len(self.st.views) == 1
+    self.st.assert_same()
+
+  def test_reshape_combining_1(self):
+    self.st = CheckingShapeTracker((2,1,10))
+    self.st.pad(((2,6), (0,0), (0,0)))
+    self.st.reshape((100,))
+    assert len(self.st.views) == 1
+    self.st.assert_same()
+
+  @unittest.skip("Can't make this optimization yet")
```

Comment:
any thoughts on how to do these and the one below?

Reply:
`test_reshape_combining_2` is a bit tricky. This is because "assumes view can be reshaped to new_shape (if it had no mask)" this assumption doesn't hold here. Here the reshape is possible because of mask else it wouldn't have been possible.

```
st = ShapeTracker((View.create(shape=(10,10), strides = (0,1)),))
st = st.reshape((100,))
print(st.expr_idxs()) # ((Variable('idx0', 0, 99)%10), 1)
```
Modulo is the problem here.
I'll give it a thought if we can somehow use masking information prior

Reply:
@chenyuxyz I tried doing this but failed. Including mask information early is more complicated than I thought and makes code much harder to reason